### DISCUSSION TITLE
Delay Effect

### DISCUSSION THREAD

**Discussion by jmtalbert** [pschatzmann/arduino-audio-tools#154] [Q&A] [answered]
The Delay Class in Effect.h seems to be just repeating the input stream.

This line within  effect_t process(effect_t input) looks wrong to me, though I'm not yet familiar with C++ templates, the ('s and <'s don't balance.

int32_t value = (p_history->available()<sampleCount) ? input : p_history->read();

**Comment by pschatzmann** [comment]
This looks right to me: if the RingBuffer is smaller then sampleCount we return the input otherwise we take the oldest sample from the RingBuffer. Maybe it gets clearer by adding a space ?

int32_t value = (p_history->available() < sampleCount) ? input : p_history->read();

**Comment by jmtalbert** [comment, marked as answer]
Found a fix that got the Delay going in the file scr/AudioTools/Buffers.h

Under class RingBuffer  you had
virtual T read(){
            if (isEmpty())
                return -1;

            uint8_t value = _aucBuffer[_iTail];
            _iTail = nextIndex(_iTail);
            _numElems--;

            return value;
        }

I replaced "uint8_t" with "T" and started getting about a one second delay.
However, the delay duration is not changing.  The depth or percent is changing fine.

**Comment by pschatzmann** [comment]
Cool, I also corrected the updateBufferSize() method to create a new Ringbuffer when the p_ms is changing (and not only increasing).
This should fix the issue with the delay duration updates...
